adala/runtimes/base.py

```diff
@@ -67,25 +74,49 @@ def batch_to_batch(
         output_template: str,
         extra_fields: Optional[Dict[str, str]] = None,
         field_schema: Optional[Dict] = None,
-        instructions_first: bool = True,
+        instructions_first: bool = True
     ) -> InternalDataFrame:
         """
         Processes a record.
+        It supports parallel processing of the batch:
+         - when the `concurrency` is set to -1 (using all available CPUs),
+         - when the `concurrency` is set to 1 (sequential processing),
+         - when the `concurrency` is set to a fixed number of CPUs.
+        Please note that parallel processing doubles the memory footprint compared to sequential processing.
 
         Args:
             batch (InternalDataFrame): The batch to process.
             input_template (str): The input template.
-            instructions_template (str): The instructions template.
+            instructions_template (str): The instructions' template.
             output_template (str): The output template.
             extra_fields (Optional[Dict[str, str]]): Extra fields to use in the templates. Defaults to None.
             field_schema (Optional[Dict]): Field JSON schema to use in the templates. Defaults to all fields are strings,
                 i.e. analogous to {"field_n": {"type": "string"}}.
             instructions_first (bool): Whether to put instructions first. Defaults to True.
-
         Returns:
             InternalDataFrame: The processed batch.
         """
-        output = batch.progress_apply(
+        if self.concurrency == -1:
+            # run batch processing each row in a parallel way, using all available CPUs
+            logger.info("Running batch processing in parallel using all available CPUs")
+            pandarallel.initialize(progress_bar=self.verbose)
+            apply_func = batch.parallel_apply
+        elif self.concurrency == 1:
+            # run batch processing each row in a sequential way
+            logger.info("Running batch processing sequentially")
+            if self.verbose:
+                apply_func = batch.progress_apply
+            else:
+                apply_func = batch.apply
+        elif self.concurrency > 1:
+            # run batch processing each row in a parallel way, using a fixed number of CPUs
+            logger.info(f"Running batch processing in parallel using {self.concurrency} CPUs")
+            pandarallel.initialize(nb_workers=self.concurrency, progress_bar=self.verbose)
```

Comment:
Would still like to get a comment here explaining the doubled memory footprint, but not PR-blocking worthy